jacodb-ets/src/main/kotlin/org/jacodb/ets/graph/EtsApplicationGraph.kt

```diff
@@ -42,13 +88,287 @@ class EtsApplicationGraphImpl(
+        }
+
+        // If the callee signature is ideal, resolve it directly:
+        if (callee.enclosingClass.isIdeal()) {
+            if (callee in cacheMethodWithIdealSignature) {
+                val resolved = cacheMethodWithIdealSignature.getValue(callee)
+                if (resolved.isSome) {
+                    stats.resolvedByTotallyMatchedCache.incrementAndGet()
+                    return sequenceOf(resolved.getOrThrow())
+                } else {
+                    stats.cachedTotalMatchAsUnknown.incrementAndGet()
+                    return emptySequence()
+                }
+            }
+
+            val cls = lookupClassWithIdealSignature(callee.enclosingClass)
+
+            val resolved = run {
+                if (cls.isNone) {
+                    emptySequence()
+                } else {
+                    cls.getOrThrow().methods.asSequence().filter { it.name == callee.name }
+                }
+            }
+            if (resolved.none()) {
+                cacheMethodWithIdealSignature[callee] = Maybe.none()
+                stats.lookupWithIdealSignatureFailed.incrementAndGet()
+                return emptySequence()
+            }
+            val r = resolved.singleOrNull()
+                ?: error("Multiple methods with the same complete signature: ${resolved.toList()}")
+            cacheMethodWithIdealSignature[callee] = Maybe.some(r)
+            stats.resolvedByIdealSignature.incrementAndGet()
+            return sequenceOf(r)
+        }
+
+        // If the callee signature is not ideal, resolve it via a partial match...
+        check(!callee.enclosingClass.isIdeal())
+
+        val cls = lookupClassWithIdealSignature(node.method.enclosingClass).let {
+            if (it.isNone) {
+                error("Could not find the enclosing class: ${node.method.enclosingClass}")
+            }
+            it.getOrThrow()
+        }
+
+        // If the complete signature match failed,
+        // try to find the unique not-the-same neighbour method in the same class:
+        val neighbors = cls.methods
+            .asSequence()
+            .filter { it.name == callee.name }
+            .filterNot { it.name == node.method.name }
+            .toList()
+        if (neighbors.isNotEmpty()) {
+            val s = neighbors.singleOrNull()
+                ?: error("Multiple methods with the same name: $neighbors")
+            cachePartiallyMatchedCallees[callee] = listOf(s)
+            stats.resolvedByNeighbour.incrementAndGet()
+            return sequenceOf(s)
+        }
+
+        // NOTE: cache lookup MUST be performed AFTER trying to match the neighbour!
+        if (callee in cachePartiallyMatchedCallees) {
+            val s = cachePartiallyMatchedCallees.getValue(callee).asSequence()
+            if (s.none())
+                stats.cachedPartialMatchAsUnknown.incrementAndGet()
+            else
+                stats.resolvedByPartiallyMatchedCache.incrementAndGet()
+            return s
```

Comment:
Missing brackets

Reply:
Fixed. Fortunately, it was not me who introduced such a deviation from our blessed code style.